src/audio/google/google_rtc_audio_processing.c

```diff
@@ -633,121 +633,139 @@ static int google_rtc_audio_processing_prepare(struct processing_module *mod,
-			++cd->aec_reference_frame_index;
-
-			if (cd->aec_reference_frame_index == cd->num_frames) {
-				GoogleRtcAudioProcessingAnalyzeRender_int16(cd->state,
-									    cd->aec_reference_buffer);
-				cd->aec_reference_frame_index = 0;
-			}
-		}
-		num_samples_remaining -= n;
-		ref = audio_stream_wrap(ref_stream, ref);
-	}
-	input_buffers[cd->aec_reference_source].consumed = num_aec_reference_bytes;
-
-	mic_streamb = &input_buffers[cd->raw_microphone_source];
-	mic_stream = mic_streamb->data;
-	out_streamb = &output_buffers[0];
-	out_stream = out_streamb->data;
-
-	src = audio_stream_get_rptr(mic_stream);
-	dst = audio_stream_get_wptr(out_stream);
-
-	frames = input_buffers[cd->raw_microphone_source].size;
-	num_frames_remaining = frames;
-
-	while (num_frames_remaining) {
-		nmax = audio_stream_frames_without_wrap(mic_stream, src);
-		n = MIN(num_frames_remaining, nmax);
-		nmax = audio_stream_frames_without_wrap(out_stream, dst);
-		n = MIN(n, nmax);
-		for (i = 0; i < n; i++) {
-			memcpy_s(&(cd->raw_mic_buffer[cd->raw_mic_buffer_frame_index *
-						      cd->num_capture_channels]),
-				 cd->num_frames * cd->num_capture_channels *
-				 sizeof(cd->raw_mic_buffer[0]), src,
-				 sizeof(int16_t) * cd->num_capture_channels);
-			++cd->raw_mic_buffer_frame_index;
-
-			memcpy_s(dst, cd->num_frames * cd->num_capture_channels *
-				 sizeof(cd->output_buffer[0]),
-				 &(cd->output_buffer[cd->output_buffer_frame_index *
-						     cd->num_capture_channels]),
-				 sizeof(int16_t) * cd->num_capture_channels);
-			++cd->output_buffer_frame_index;
-
-			if (cd->raw_mic_buffer_frame_index == cd->num_frames) {
-				GoogleRtcAudioProcessingProcessCapture_int16(cd->state,
-									     cd->raw_mic_buffer,
-									     cd->output_buffer);
-				cd->output_buffer_frame_index = 0;
-				cd->raw_mic_buffer_frame_index = 0;
-			}
+	struct comp_buffer *refbuf = container_of(ref, struct comp_buffer, stream);
+	bool ref_ok = refbuf->source->state == COMP_STATE_ACTIVE;
 
-			src += audio_stream_get_channels(mic_stream);
-			dst += audio_stream_get_channels(out_stream);
+	/* Would be cleaner to store a bit of state to elide a bzero
+	 * we already did, but we'd be doing the copy of real data in
+	 * the ref_ok state anyway.
+	 */
+	if (!ref_ok)
+		bzero(cd->aec_reference_buffer,
+		      (cd->num_frames * cd->num_aec_reference_channels
+		       * sizeof(cd->aec_reference_buffer[0])));
+
+	int micchan = audio_stream_get_channels(mic);
+	int refchan = audio_stream_get_channels(ref);
+
+	int fmic = audio_stream_get_avail_frames(mic);
+	int fref = audio_stream_get_avail_frames(ref);
+	int frames = ref_ok ? MIN(fmic, fref) : fmic;
+	int n, frames_rem;
+
+	/* If fref > fmic (common at pipeline startup if
+	 * playback was already active), we should consume the early
+	 * samples so AEC compares the most recent values.
+	 */
+	if (ref_ok && fref > fmic)
+		audio_stream_consume(ref, (fref - fmic) * audio_stream_frame_bytes(ref));
```

Comment:
In an offline comment, Per points out that this will drop playback samples in the middle of the stream in the case that the buffer updates aren't always deterministic and synchronous (something that the code has always assumed, but will probably NOT be the case in the DP future).  This probably need to be based on a one-shot flag so that we do it only once after each ref_ok false->true transition.